File: python/lexer.py
from enum import Enum

# exception handling
class Error(Exception):
    def __init__(self, error_code=None, token=None, message=None):
        self.error_code = error_code
        self.token = token
        self.message = f'{self.__class__.__name__}: {message}'

class LexerError(Error):
    pass

class TokenType(Enum):
    # single character token types
    PLUS        = '+'
    MINUS       = '-'
    MUL         = '*'
    FLOAT_DIV   = '/'
    LPAREN      = '('
    RPAREN      = ')'
    LBRACE      = '{'
    RBRACE      = '}'
    SEMI        = ';'
    DOT         = '.'
    COLON       = ':'
    COMMA       = ','
    ASSIGN      = ':='
    LESS        = '<'
    LESS_EQUAL      = '<='
    GREATER         = '>'
    GREATER_EQUAL   = '>='
    BANG            = '!'
    EQUAL           = '='
    BANG_EQUAL      = '!='
    # block of reserved words
    PROGRAM     = 'PROGRAM'
    INTEGER     = 'INTEGER'
    REAL        = 'REAL'
    INTEGER_DIV = 'DIV'
    VAR         = 'VAR'
    PROCEDURE   = 'PROCEDURE'
    BEGIN       = 'BEGIN'
    END         = 'END'
    # misc
    IDENT           = 'IDENT'
    INTEGER_CONST   = 'INTEGER_CONST'
    REAL_CONST      = 'REAL_CONST'
    EOF             = 'EOF'

class Token(object):
    def __init__(self, type, value, line = 0, col = 0):
        self.type = type
        self.value = value
        self.line = line
        self.col = col

class Lexer(object):
    def __init__(self, text):
        self.text = text
        self.pos = 0
        self.current_char = self.text[self.pos]
        self.lineno = 1
        self.column = 1
        self.reserved_keywords = {
            'PROGRAM': TokenType.PROGRAM,
            'INTEGER': TokenType.INTEGER,
            'REAL': TokenType.REAL,
            'DIV': TokenType.INTEGER_DIV,
            'VAR': TokenType.VAR,
            'PROCEDURE': TokenType.PROCEDURE,
            'BEGIN': TokenType.BEGIN,
            'END': TokenType.END,
        }

    def error(self):
        s = "Lexer error on '{lexeme}' line: {lineno} column: {column}".format(
            lexeme=self.current_char,
            lineno=self.lineno,
            column=self.column,
        )
        raise Exception(s)

    def advance(self):
        self.pos += 1
        if self.pos > len(self.text) - 1:
            self.current_char = None
        else:
            self.current_char = self.text[self.pos]
        # new line
        if self.current_char == '\n':
            self.lineno += 1
            self.column = 1
    
    def peek(self):
        peek_pos = self.pos + 1
        if peek_pos > len(self.text) - 1:
            return None
        else:
            return self.text[peek_pos]
    
    def skip_whitespace(self):
        while self.current_char != None and self.current_char.isspace():
            self.advance()
    
    def skip_comments(self):
        while self.current_char != None and self.current_char != '}':
            self.advance()

        if self.current_char == None:
            self.error()
        self.advance() # eat '}'
    
    def number(self):
        result = ''
        while self.current_char != None and self.current_char.isdigit():
            result += self.current_char
            self.advance()
        # decimal number
        if self.current_char == '.' and self.peek().isdigit():
            result += self.current_char
            self.advance() # skip the '.'
            while self.current_char != None and self.current_char.isdigit():
                result += self.current_char
                self.advance()
            return self.new_token(TokenType.REAL_CONST, float(result))
        else:
            return self.new_token(TokenType.INTEGER_CONST, int(result))
    
    def isalpha(self, ch):
        return ch.isalnum() or ch == '_'

    def identifier(self):
        result = ''
        while self.current_char != None and self.current_char.isalpha():
            result += self.current_char
            self.advance()
        token_type = self.reserved_keywords.get(result, TokenType.IDENT)
        return self.new_token(token_type, result)
    
    def new_token(self, type = None, value = None):
        if value == None:
            value = type
        return Token(type, value, self.lineno, self.column)

    def get_next_token(self):
        while self.current_char != None:
            if self.current_char.isspace():
                self.skip_whitespace()
                continue

            if self.current_char == '{':
                self.skip_comments()
                continue

            if self.current_char.isdigit():
                return self.number()
            
            if self.isalpha(self.current_char):
                return self.identifier()

            if self.current_char == '+':
                self.advance()
                return self.new_token(TokenType.PLUS)

            if self.current_char == '-':
                self.advance()
                return self.new_token(TokenType.MINUS)

            if self.current_char == '*':
                self.advance()
                return self.new_token(TokenType.MUL)
            
            if self.current_char == '/':
                self.advance()
                return self.new_token(TokenType.FLOAT_DIV)
            
            if self.current_char == '(':
                self.advance()
                return self.new_token(TokenType.LPAREN)
            
            if self.current_char == ')':
                self.advance()
                return self.new_token(TokenType.RPAREN)

            if self.current_char == ';':
                self.advance()
                return self.new_token(TokenType.SEMI)

            if self.current_char == '.':
                self.advance()
                return self.new_token(TokenType.DOT)

            if self.current_char == ':':
                if self.peek() == '=':
                    self.advance() # eat ':'
                    self.advance() # eat '='
                    return self.new_token(TokenType.ASSIGN)
                else:
                    self.advance()
                    return self.new_token(TokenType.COLON)

            if self.current_char == ',':
                self.advance()
                return self.new_token(TokenType.COMMA)
            
            if self.current_char == '<':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    return self.new_token(TokenType.LESS_EQUAL)
                else:
                    return self.new_token(TokenType.LESS)
            
            if self.current_char == '>':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    return self.new_token(TokenType.GREATER_EQUAL)
                else:
                    return self.new_token(TokenType.GREATER)
            
            if self.current_char == '!':
                self.advance()
                if self.current_char == '=':
                    self.advance()
                    return self.new_token(TokenType.BANG_EQUAL)
                else:
                    return self.new_token(TokenType.BANG)

            self.error()

        return Token(TokenType.EOF, None)

def main():
    # while True:
    #     try:
    #         text = input('lexer> ')
    #     except EOFError:
    #         break
    #     if not text:
    #         continue
    text = """\
program Main;

procedure Alpha(a : integer; b : integer);
var x : integer;

   procedure Beta(a : integer; b : integer);
   var x : integer;
   begin
      x := a * 10 + b * 2;
   end;

begin
   x := (a + b ) * 2;

   Beta(5, 10);      { procedure call }
end;

begin { Main }

   Alpha(3 + 5, 7);  { procedure call }

end.  { Main }
        """
    lexer = Lexer(text)
    token = lexer.get_next_token()
    while token.type != TokenType.EOF:
        print_token(token)
        token = lexer.get_next_token()
    print_token(token)
        
def print_token(token):
    print('type: {type} value: {value}'.format(
        type = token.type,
        value = token.value
    ))

if __name__ == '__main__':
    main()

File: go/lexer.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
)

type TokenType string

const (
	// single character token types
	PLUS      = "+"
	MINUS     = "-"
	MUL       = "*"
	FLOAT_DIV = "/"
	LPAREN    = "("
	RPAREN    = ")"
	LBRACE    = "{"
	RBRACE    = "}"
	SEMI      = ";"
	DOT       = "."
	COLON     = ":"
	COMMA     = ","
	ASSIGN    = ":="
	LESS      = "<"
	// block of reserved words
	PROGRAM     = "PROGRAM"
	INTEGER     = "INTEGER"
	REAL        = "REAL"
	INTEGER_DIV = "DIV"
	VAR         = "VAR"
	PROCEDURE   = "PROCEDURE"
	BEGIN       = "BEGIN"
	END         = "END"
	// misc
	IDENT         = "IDENT"
	INTEGER_CONST = "INTEGER_CONST"
	REAL_CONST    = "REAL_CONST"
	EOF           = "EOF"
	NONE          = 255
)

// Reserved keywords
var keywords = map[string]TokenType{
	"PROGRAM":   PROGRAM,
	"INTEGER":   INTEGER,
	"REAL":      REAL,
	"DIV":       INTEGER_DIV,
	"VAR":       VAR,
	"PROCEDURE": PROCEDURE,
	"BEGIN":     BEGIN,
	"END":       END,
}

type Token struct {
	Type   TokenType
	Value  interface{}
	Line   int
	Column int
}

type Lexer struct {
	Text        string
	CurrentChar byte
	Pos         int
	Line        int
	Column      int
}

func NewLexer(text string) *Lexer {
	l := Lexer{Text: text}
	l.Pos = 0
	l.CurrentChar = l.Text[l.Pos]
	l.Column = 1
	l.Line = 1
	return &l
}

func (l *Lexer) LookupIdent(ident string) TokenType {
	if tok, ok := keywords[ident]; ok {
		return tok
	}
	return IDENT
}

func (l *Lexer) error() {
	fmt.Printf("Unexpected character '%s'\n", string(l.CurrentChar))
	os.Exit(1)
}

func (l *Lexer) advance() {
	if l.CurrentChar == '\r' { // for Windows compatibility
		l.nextChar()
	}
	if l.CurrentChar == '\n' {
		l.Line++
		l.Column = 0
		l.nextChar()
	}
	l.nextChar()
}

func (l *Lexer) nextChar() {
	l.Pos++
	if l.Pos > len(l.Text)-1 {
		l.CurrentChar = NONE
	} else {
		l.CurrentChar = l.Text[l.Pos]
	}
}

func (l *Lexer) peek() byte {
	peekPos := l.Pos + 1
	if peekPos > len(l.Text)-1 {
		return NONE
	} else {
		return l.Text[peekPos]
	}
}

func (l *Lexer) isSpace(ch byte) bool {
	return ch == ' '
}

func (l *Lexer) isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

func (l *Lexer) isLetter(ch byte) bool {
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_'
}

func (l *Lexer) skipWhiteSpace() {
	for l.CurrentChar != NONE && l.isSpace(l.CurrentChar) {
		l.advance()
	}
}

func (l *Lexer) number() Token {
	result := ""
	for l.CurrentChar != NONE && l.isDigit(l.CurrentChar) {
		result += string(l.CurrentChar)
		l.advance()
	}
	if l.CurrentChar == '.' && l.isDigit(l.peek()) {
		result += string(l.CurrentChar)
		l.advance() // eat '.'
		for l.CurrentChar != NONE && l.isDigit(l.CurrentChar) {
			result += string(l.CurrentChar)
			l.advance()
		}
		real, _ := strconv.ParseFloat(result, 64)
		return Token{Type: REAL_CONST, Value: real, Line: l.Line, Column: l.Column}
	}
	integer, _ := strconv.Atoi(result)
	return Token{Type: INTEGER_CONST, Value: integer, Line: l.Line, Column: l.Column}
}

func (l *Lexer) identifier() Token {
	result := ""
	for l.CurrentChar != NONE && l.isLetter(l.CurrentChar) {
		result += string(l.CurrentChar)
		l.advance()
	}
	return Token{Type: l.LookupIdent(result), Value: result, Line: l.Line, Column: l.Column}
}

func (l *Lexer) skipComments() {
	for l.CurrentChar != NONE && l.CurrentChar != '}' {
		l.advance()
	}
	if l.CurrentChar == NONE {
		fmt.Print("Unexpected End Of File.")
		os.Exit(1)
	}
	l.advance() // eat closing comment '}'
}

func (l *Lexer) GetNextToken() Token {
	for l.CurrentChar != NONE {
		if l.isSpace(l.CurrentChar) {
			l.skipWhiteSpace()
			continue
		}
		if l.CurrentChar == '{' {
			l.skipComments()
			continue
		}
		if l.isDigit(l.CurrentChar) {
			return l.number()
		}
		if l.isLetter(l.CurrentChar) {
			return l.identifier()
		}
		if l.CurrentChar == '+' {
			l.advance()
			return Token{Type: PLUS, Value: "+", Line: l.Line, Column: l.Column}
		}
		if l.CurrentChar == '-' {
			l.advance()
			return Token{Type: MINUS, Value: "-", Line: l.Line, Column: l.Column}
		}
		if l.CurrentChar == '*' {
			l.advance()
			return Token{Type: MUL, Value: "*", Line: l.Line, Column: l.Column}
		}
		if l.CurrentChar == '/' {
			l.advance()
			return Token{Type: FLOAT_DIV, Value: "/", Line: l.Line, Column: l.Column}
		}
		if l.CurrentChar == '(' {
			l.advance()
			return Token{Type: LPAREN, Value: "(", Line: l.Line, Column: l.Column}
		}
		if l.CurrentChar == ')' {
			l.advance()
			return Token{Type: RPAREN, Value: ")", Line: l.Line, Column: l.Column}
		}
		if l.CurrentChar == '.' {
			l.advance()
			return Token{Type: DOT, Value: ".", Line: l.Line, Column: l.Column}
		}
		if l.CurrentChar == ',' {
			l.advance()
			return Token{Type: COMMA, Value: ",", Line: l.Line, Column: l.Column}
		}
		if l.CurrentChar == ';' {
			l.advance()
			return Token{Type: SEMI, Value: ";", Line: l.Line, Column: l.Column}
		}
		if l.CurrentChar == ':' {
			if l.peek() == '=' {
				l.advance()
				l.advance()
				return Token{Type: ASSIGN, Value: ":=", Line: l.Line, Column: l.Column}
			}
			l.advance()
			return Token{Type: COLON, Value: ":", Line: l.Line, Column: l.Column}
		}
		if l.CurrentChar == '\r' || l.CurrentChar == '\n' {
			l.advance()
		} else {
			l.error()
		}
	}
	return Token{Type: EOF, Value: NONE}
}

func main() {
	//text := "program Main;procedure Alpha(a : integer; b : integer); var x : integer;"

	pwd, _ := os.Getwd()
	txt, _ := ioutil.ReadFile(pwd + "/test.txt")
	content := string(txt)

	lexer := NewLexer(content)
	token := lexer.GetNextToken()
	for token.Value != NONE {
		fmt.Printf("%-v\n", token)
		token = lexer.GetNextToken()
	}
}
